sklearn/cluster/hierarchical.py

```diff
@@ -693,14 +695,10 @@ def fit(self, X, y=None):
         self
         """
         X = check_array(X, ensure_min_samples=2, estimator=self)
-        memory = self.memory
-        if memory is None:
-            memory = Memory(cachedir=None, verbose=0)
-        elif isinstance(memory, six.string_types):
-            memory = Memory(cachedir=memory, verbose=0)
-        elif not isinstance(memory, Memory):
+        memory = check_memory(self.memory)
+        if not hasattr(memory, 'cache'):
```

Comment:
This raise statement will be duplicate. You need to include it in `check_memory`.